tests/components/zha/test_discover.py

```diff
@@ -789,18 +818,280 @@ async def test_quirks_v2_entity_discovery_errors(
+
+    metadata_type = type(entity_metadata)
+    if metadata_type in DEVICE_CLASS_TYPES:
+        device_class = entity_metadata.device_class
+        if device_class is not None and translation_key is not None:
+            m1 = "translation_key and device_class are both set - translation_key: "
+            m2 = f"{translation_key} device_class: {device_class} for {platform.name} "
+            raise ValueError(f"{m1}{m2}{quirk}")
+
+
+def validate_metadata(validator: Callable) -> None:
+    """Ensure v2 quirks metadata does not violate HA rules."""
+    all_v2_quirks = itertools.chain.from_iterable(
+        zigpy.quirks._DEVICE_REGISTRY._registry_v2.values()
+    )
+    translations = load_json("homeassistant/components/zha/strings.json")
+    for quirk in all_v2_quirks:
+        for entity_metadata in quirk.entity_metadata:
+            platform = Platform(entity_metadata.entity_platform.value)
+            validator(quirk, entity_metadata, platform, translations)
+
+
+def bad_translation_key(v2_quirk: QuirksV2RegistryEntry) -> QuirksV2RegistryEntry:
+    """Introduce a bad translation key."""
+    return v2_quirk.sensor(
+        zigpy.zcl.clusters.general.OnOff.AttributeDefs.off_wait_time.name,
+        zigpy.zcl.clusters.general.OnOff.cluster_id,
+        entity_type=EntityType.CONFIG,
+        translation_key="missing_translation_key",
+    )
+
+
+def bad_device_class_unit_combination(
+    v2_quirk: QuirksV2RegistryEntry,
+) -> QuirksV2RegistryEntry:
+    """Introduce a bad device class and unit combination."""
+    return v2_quirk.sensor(
+        zigpy.zcl.clusters.general.OnOff.AttributeDefs.off_wait_time.name,
+        zigpy.zcl.clusters.general.OnOff.cluster_id,
+        entity_type=EntityType.CONFIG,
+        unit="invalid",
+        device_class="invalid",
+        translation_key="analog_input",
+    )
+
+
+def bad_device_class_translation_key_usage(
+    v2_quirk: QuirksV2RegistryEntry,
+) -> QuirksV2RegistryEntry:
+    """Introduce a bad device class and translation key combination."""
+    return v2_quirk.sensor(
+        zigpy.zcl.clusters.general.OnOff.AttributeDefs.off_wait_time.name,
+        zigpy.zcl.clusters.general.OnOff.cluster_id,
+        entity_type=EntityType.CONFIG,
+        translation_key="invalid",
+        device_class="invalid",
+    )
+
+
+@pytest.mark.parametrize(
+    ("augment_method", "validate_method", "expected_exception_string"),
+    [
+        (
+            bad_translation_key,
+            validate_translation_keys,
+            "Missing translation key: missing_translation_key",
+        ),
+        (
+            bad_device_class_unit_combination,
+            validate_device_class_unit,
+            "cannot have both unit and device_class",
+        ),
+        (
+            bad_device_class_translation_key_usage,
+            validate_translation_keys_device_class,
+            "cannot have both a translation_key and a device_class",
```

Comment:
I'm missing something: `validate_device_class_unit` would throw "translation_key and device_class are both set"
and `validate_translation_keys_device_class` would throw "translation_key and device_class are both set", so how come we can (successfully?) expect another exception string here?